qualtran/_infra/adjoint.py

```diff
@@ -0,0 +1,168 @@
+    bb, _ = BloqBuilder.from_signature(new_signature)
+    for binst, preds, succs in bloqnections:
+        # Instead of get_me returning the right element of a predecessor connection,
+        # it's the left element of a successor connection.
+        soqs = _cxn_to_soq_dict(
+            binst.bloq.signature.rights(),
+            succs,
+            get_me=lambda x: x.left,
+            get_assign=lambda x: x.right,
+        )
+        soqs = _map_soqs(soqs, soq_map)
+
+        old_o_soqs = tuple(_reg_to_soq(binst, reg) for reg in binst.bloq.signature.lefts())
+        new_o_soqs = bb.add_t(binst.bloq.adjoint(), **soqs)
+        soq_map.extend(zip(old_o_soqs, new_o_soqs))
+
+    # Instead of finalizing with RightDangle predecessors, we use LeftDangle successors
+    fsoqs = _map_soqs(_adjoint_final_soqs(cbloq, new_signature), soq_map)
+    return bb.finalize(**fsoqs)
+
+
+@frozen
+class Adjoint(GateWithRegisters):
+    """The standard adjoint of `subbloq`.
+
+    This metabloq generally delegates all of its protocols (with modifications, read on) to
+    `subbloq`. This class is used in the default implementation of the adjoint protocol, i.e.,
+    in the default implementation of `Bloq.adjoint()`.
+
+    This metabloq is appropriate in most cases since there rarely a specialized
+    (one level) decomposition for a bloq's adjoint. Exceptions can be found for decomposing
+    some low-level primitives, for example `And`. Even if you use bloqs with specialized
+    adjoints in your decomposition (i.e. you use `And`), you can still rely on this standard
+    behavior.
+
+    This bloq is defined entirely in terms of how it delegates its protocols. The following
+    protocols delegate to `subbloq` (with appropriate modification):
+
+     - **Signature**: The signature is the adjoint of `subbloqs`'s signature. Namely, LEFT
+       and RIGHT registers are swapped.
+     - **Decomposition**: The decomposition is the adjoint of `subbloq`'s decomposition. Namely,
+       the order of operations in the resultant `CompositeBloq` is reversed and each bloq is
+       replaced with its adjoint.
+     - **Adjoint**: The adjoint of an `Adjoint` bloq is the subbloq itself.
+     - **Call graph**: The call graph is the subbloq's call graph, but each bloq is replaced
+       with its adjoint.
+     - **Cirq Interop**: The default `Bloq` implementation is used, which goes via `BloqAsCirqGate`
+       as usual.
+     - **Wire Symbol**: The wire symbols are the adjoint of `subbloq`'s wire symbols. Namely,
+       left- and right-oriented symbols are flipped.
+     - **Names**: The string names / labels are that of the `subbloq` with a dagger symbol appended.
+
+    Some protocols are impossible to delegate specialized implementations. The `Adjoint` bloq
+    supports the following protocols with "decompose-only" implementations. This means we always
+    go via the bloq's decomposition instead of preferring specialized implementations provided by
```

Comment:
To clarify, we go by the Adjoint bloq's decomposition right? Can we add some tests for the adjoint tensor simulations?

Reply:
`Adjoint` doesn't override any of the tensor methods, so as long as the decomposition is correct and the tensors-from-decomposition code is correct there's nothing new to test.

I think tensor tests will be merited and valuable when we have some bloqs that support the adjoint protocol #566 . Right now: you'd keep decomposing and wrapping in `Adjoint(...)` until you get to atomic bloqs and an exception will be raised. You need `MyBloq.adjoint()` to sometimes return concrete leaf bloqs